File: src/csharp/Foundgine.Core/Semantic/Metadata/AuthorizationMetadata.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>
/// A compile-time authorization predicate attached to a semantic connection.
/// The expression is descriptive metadata; providers may later lower the
/// predicate into their native authorization/filter representation.
/// </summary>
public sealed record AuthorizationMetadata(
    AuthorizationId Id,
    ConnectionId ConnectionId,
    string Name,
    string SourceMember,
    Type ContextType,
    Type ResourceType,
    string Expression,
    AuthorizationPredicate? Predicate = null);

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/AuthorizationMetadata.java
package com.foundgine.core.semantic.metadata;

import com.foundgine.core.abstractions.AuthorizationId;
import com.foundgine.core.abstractions.AuthorizationPredicate;
import com.foundgine.core.abstractions.ConnectionId;

public record AuthorizationMetadata(AuthorizationId id, ConnectionId connectionId, String name, String sourceMember,
		Class<?> contextType, Class<?> resourceType, String expression, AuthorizationPredicate predicate) {
	public AuthorizationMetadata(AuthorizationId id, ConnectionId connectionId, String name, String sourceMember,
			Class<?> contextType, Class<?> resourceType, String expression) {
		this(id, connectionId, name, sourceMember, contextType, resourceType, expression, null);
	}
}
